ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/versions/bellatrix/block/BlockProcessorBellatrix.java

```diff
@@ -79,68 +80,72 @@ public void processBlock(
+    } else {
+      final ExecutionPayload executionPayload =
+          blockBody.getOptionalExecutionPayload().orElseThrow();
+      executionPayloadHeader =
+          schemaDefinitions
+              .getExecutionPayloadHeaderSchema()
+              .createFromExecutionPayload(executionPayload);
+    }
+
     processBlockHeader(state, block);
     if (miscHelpersBellatrix.isExecutionEnabled(genericState, block)) {
-      processExecutionPayload(state, blockBody.getExecutionPayload(), payloadExecutor);
+      processExecutionPayload(
+          state, executionPayloadHeader, blockBody.getOptionalExecutionPayload(), payloadExecutor);
     }
     processRandaoNoValidation(state, block.getBody());
     processEth1Data(state, block.getBody());
     processOperationsNoValidation(state, block.getBody(), indexedAttestationCache);
-    processSyncAggregate(state, blockBody.getSyncAggregate(), signatureVerifier);
+    processSyncAggregate(
+        state, blockBody.getOptionalSyncAggregate().orElseThrow(), signatureVerifier);
   }
 
   @Override
   public void processExecutionPayload(
       final MutableBeaconState genericState,
-      final ExecutionPayload payload,
+      final ExecutionPayloadHeader executionPayloadHeader,
+      final Optional<ExecutionPayload> executionPayload,
       final OptimisticExecutionPayloadExecutor payloadExecutor)
       throws BlockProcessingException {
     final MutableBeaconStateBellatrix state = MutableBeaconStateBellatrix.required(genericState);
     if (miscHelpersBellatrix.isMergeTransitionComplete(state)) {
-      if (!payload.getParentHash().equals(state.getLatestExecutionPayloadHeader().getBlockHash())) {
+      if (!executionPayloadHeader
+          .getParentHash()
+          .equals(state.getLatestExecutionPayloadHeader().getBlockHash())) {
         throw new BlockProcessingException(
             "Execution payload parent hash does not match previous execution payload header");
       }
     }
 
     if (!beaconStateAccessors
         .getRandaoMix(state, beaconStateAccessors.getCurrentEpoch(state))
-        .equals(payload.getPrevRandao())) {
+        .equals(executionPayloadHeader.getPrevRandao())) {
       throw new BlockProcessingException("Execution payload random does not match state randao");
     }
 
     if (!miscHelpersBellatrix
         .computeTimeAtSlot(state, state.getSlot())
-        .equals(payload.getTimestamp())) {
+        .equals(executionPayloadHeader.getTimestamp())) {
       throw new BlockProcessingException(
           "Execution payload timestamp does not match time for state slot");
     }
 
-    final boolean optimisticallyAccept =
-        payloadExecutor.optimisticallyExecute(state.getLatestExecutionPayloadHeader(), payload);
-    if (!optimisticallyAccept) {
-      throw new BlockProcessingException("Execution payload was not optimistically accepted");
+    if (executionPayload.isPresent()) {
```

Comment:
This makes me a bit nervous that we may wind up not having the payload here when importing the block and this would then result in it being considered valid. Not entirely sure how we resolve that - maybe a check in `ForkChoice.onBlock` to require the block not be blinded could be enough?

Reply:
right.

> maybe a check in ForkChoice.onBlock to require the block not be blinded could be enough?

should work, but I'll think about an alternative.

Reply:
I think this needs to be more structural. `BlockProcessor` requires a bit more of context so I made the `PayloadExecutor` to be optional and removed the NOOP. This way we are more strict in communicating when we have to execute a payload or not. And if we have to, we require the full payload to be there.

WDYT?